opendatakit/javarosa
# Issue 534 fix date utils test

Closes #534

This PR has become more than simply fixing the DateUtilsTest class into a review of the whole test suite to ensure we always deal with timezones and locales in a way that's consistent, safe and easier to understand.

#### What has been done to verify that this works as intended?
- Run automated tests
- Use [external](https://www.epochconverter.com/timezones?q=1580378400000&tz=Pacific%2FHonolulu) tools to double-check that the `java.time` values used to replace long timestamps actually represent the same instant.

#### Why is this the best possible solution? Were any other approaches considered?
- Using `java.time` as much as possible is the safest way to move forward even if the implementation codebase will use Jodatime while we support Android API levels that don't include it.

  This is the standard way to deal with dates in Java starting from Java8 and it's very well documented with lots of examples.

  Also, using a `java.time` serves as a safe and solid baseline that provides an extra sanity check over our code and tests.

- Using human readable date and time definitions in our tests makes them easier to understand. When something goes wrong, it will be easier to detect problems if we can read the values involved, as opposed to using timestamps.

- Although Using the `@Before` and `@After` hooks to prevent side-effects due to setting default locales and time zones in test methods is super safe, using the methods provided by `SystemHelper` (new in this PR) make more explicit where this is actually required.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This PR doesn't change the implementation codebase, so no behavior change expected whatsoever.

#### Do we need any specific form for testing your changes? If so, please attach one.
No

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No
